### PR TITLE
Fixes `react-native link` for Windows

### PR DESCRIPTION
**Pull request by rozele**
Improve resilience of `link` command on Windows

There were various bugs related to `react-native link` for `react-native-windows`:    - NPM projects with multiple .csproj files could link the wrong projects
- Recent addition of ReactNativeHost (in lieu of ReactPage) was not supported
- Project GUIDs were not correctly referenced when linking dependencies
- NPM dependency projects needed to unnecessarily publish a Visual Studio solution file (i.e., .sln)
- Linking a project in a solution did not set up a default build configuration

This change set fixes all of the above. There are some remaining things to do for fit and finish:
- Support linking projects as well as their source code project (i.e., .csproj) dependencies to the RNW app solution

Fixes Microsoft/react-native-windows#738
Fixes Microsoft/react-native-windows#1576
Fixes Microsoft/react-native-windows#1578

## Motivation

The `react-native link` command is broken for version 0.50+ of `react-native-windows`.

## Test Plan

* Start localhost NPM registry (i.e., sinopia)
* Apply patch to `facebook/react-native` repo
* Publish patch to localhost registry
* Create a new project with `react-native init` from patched version of `react-native`
* Install a package with windows support (e.g., `react-native-fs`)
* Run `react-native link`
* Confirm link and unlink commands are successful for Windows.

## Related PRs

(If this PR adds or changes functionality, please take some time to update the docs at https://github.com/facebook/react-native-website, and link to your PR here.)

## Release Notes

[WINDOWS][BUGFIX][CLI] - Fix the link and unlink commands for `react-native-windows`

### CODE REVIEW

**Comment by rozele** [comment]
cc @grabbou. LMK if there are any problems with this PR. It would be great to get this patched into 0.50+

File: local-cli/link/windows/patches/makeProjectPatch.js [hunk 1]
```diff
@@ -1,14 +1,13 @@
...
+      <Project>{${windowsConfig.projectGUID}}</Project>
       <Name>${windowsConfig.projectName}</Name>
     </ProjectReference>
     `;
 
   return {
     pattern: '<ProjectReference Include="..\\..\\node_modules\\react-native-windows\\ReactWindows\\ReactNative\\ReactNative.csproj">',
     patch: projectInsert,
-    unpatch: new RegExp(`<ProjectReference.+\\s+.+\\s+.+${windowsConfig.projectName}.+\\s+.+\\s`),
   };
```
Comment (rozele) [inline local-cli/link/windows/patches/makeProjectPatch.js]: Make this more resilient to whitespace differences.

**Comment by rozele** [comment]
>   "name": "react-native",

Don't commit this file.

---
Refers to: package-lock.json:2 in 60a28d3. [](commit_id = 60a28d38f9a8be6a0c68f0ad88152bbd2046cb27, deletion_comment = False)

**Comment by rozele** [comment]
bump @grabbou - I think *fingers crossed* I finally got the eslint-bot to stop complaining :-)

**Comment by grabbou** [comment]
Hey @rozele, thanks for submitting the PR. I am going through it right now, will follow up on Slack with any questions.

I think it makes sense to cherry-pick it to previous versions (if there are documented blockers). 

On the other hand, I am wondering (which is probably not in scope for this PR), is there a way to make this `windows` folder dynamically loaded so that you can extend `link` by having code within `react-native-windows` repo?

**Comment by rozele** [comment]
@grabbou - good question. A follow up question for you would be - do you think it could be safe to execute multiple react-native CLI plugins with the same name? E.g., if I exposed the `link` command in `react-native-windows`, is it reasonable that calling `react-native link` would execute both the `react-native` implementation of link and the `react-native-windows` implementation?

If so, I think that's definitely something we could do.

**Comment by rozele** [comment]
@grabbou One potential blocker to this, however, is the configuration step for react-native CLI plugins.  Currently there is no step to check other modules for hooks into the configuration. We would also need to expose those hooks.

**Comment by rozele** [comment]
@grabbou another major blocker is the prelink and postlink assumptions for the `react-native link` command. I believe the assumption is that prelink is run exactly once before iOS, Android, and Windows packages are linked, and postlink is run exactly once after iOS, Android, and Windows are linked. If I created my own link command, it would have to run pre- and postlink twice.

**Comment by rozele** [comment]
Closing in favor of opening the CLI configuration and link command to plugin platforms (so code can be maintained in separate repos)

**Comment by grabbou** [comment]
> if I exposed the link command in react-native-windows, is it reasonable that calling react-native link would execute both the react-native implementation of link and the react-native-windows implementation?

I would assume it just overwrites as that's the default behaviour right now - could be done in a way that defining with the same name is "extending" which actually makes a lot of sense.

**Comment by grabbou** [comment]
@rozele great that my suggestion lead to a PR that actually reduces the codebase. I think at some point we might want to document that in the documentation so that others can also benefit from this great integration.

Meanwhile, great job! Moving onto another PR.

**Comment by raouldandresy** [comment]
Sorry, i just copy and paste your commit change in my react native but i get this err : 
\node_modules\react-native\local-cli\core\index.js:65
reduce((acc, pathToPlatforms) => {
^

TypeError: Reduce of empty array with no initial value
    at Array.reduce (<anonymous>)
    at Object.<anonymous> (C:/Users/*****/android/MusicProject/node_modules/react-native/local-cli/core/index.js:65:4)
    at Module._compile (module.js:649:30)
    at loader (C:\Users\*****\android\MusicProject\node_modules\babel-register\lib\node.js:144:5)
    at Object.require.extensions.(anonymous function) [as .js] (C:\Users\******\android\MusicProject\node_modules\babel-register\lib\node.js:154:7)
    at Module.load (module.js:561:32)
    at tryModuleLoad (module.js:501:12)
    at Function.Module._load (module.js:493:3)
    at Module.require (module.js:593:17)
    at require (internal/module.js:11:18)

**Comment by grabbou** [comment]
Why did you copied those changes to your code in first place?

**Pull request closed without merging**